reactors/1D_packbed.ipynb

```diff
@@ -0,0 +1,518 @@
+    "    coverages = surf.coverages  # surface site coverages\n",
+    "    sdot_g = surf.get_net_production_rates(\n",
+    "        \"gas\"\n",
+    "    )  # heterogeneous production rate of gas species\n",
+    "    sdot_s = surf.get_net_production_rates(\n",
+    "        \"Ru_surface\"\n",
+    "    )  # heterogeneous production rate of surface species\n",
+    "    wdot_g = np.zeros(n_gas)\n",
+    "    cp = gas.cp_mass  # specific heat of the mixture\n",
+    "    hk_g = gas.partial_molar_enthalpies  # partial enthalpies of gas species\n",
+    "\n",
+    "    if n_gas_reactions > 0:\n",
+    "        wdot_g = gas.net_production_rates  # homogeneous production rate of gas species\n",
+    "    mu = gas.viscosity  # viscosity of the gas-phase\n",
+    "\n",
+    "    # Calculate density using equation of state\n",
+    "    rho = gas.density\n",
+    "\n",
+    "    # Calculate flux term through the membrane\n",
+    "    memsp_pres = (\n",
+    "        p * gas.X[membrane_sp_ind]\n",
+    "    )  # partial pressure of membrane-permeable species\n",
+    "    membrane_flux = (\n",
+    "        -permeance * (memsp_pres - p_sweep) / W_g[membrane_sp_ind]\n",
+    "    )  # negative sign indicates the flux going out\n",
+    "\n",
+    "    # Conservation of total-mass\n",
+    "    sum_continuity = As * np.sum(sdot_g * W_g) + phi * np.sum(\n",
+    "        wdot_g * W_g\n",
+    "    )  # temporary variable\n",
+    "    res[offset_rhou] = (\n",
+    "        drhoudz - sum_continuity - area2vol * membrane_flux * membrane_present\n",
+    "    )\n",
+    "\n",
+    "    # Conservation of gas-phase species\n",
+    "    for k in range(n_gas):\n",
+    "        res[offset_Y + k] = (\n",
+    "            dYdz[k]\n",
+    "            + (\n",
+    "                Y[k] * sum_continuity\n",
+    "                - phi * wdot_g[k] * W_g[k]\n",
+    "                - As * sdot_g[k] * W_g[k]\n",
+    "            )\n",
+    "            / rhou\n",
+    "        )\n",
+    "        if k == membrane_sp_ind:\n",
+    "            res[offset_Y + k] -= area2vol * membrane_flux * membrane_present\n",
+    "\n",
+    "    # Conservation of site fractions (algebraic contraints in this example)\n",
+    "    # for k in range(n_surf):\n",
+    "    res[offset_Z : offset_Z + n_surf] = sdot_s\n",
+    "\n",
+    "    # For the species with largest site coverage, use the constraint sum(Zk) = 1\n",
+    "    ind_large = np.argmax(coverages)\n",
+    "    res[offset_Z + ind_large] = 1 - np.sum(coverages)\n",
+    "\n",
+    "    # Conservation of momentum\n",
+    "    u = rhou / rho\n",
+    "    res[offset_p] = dpdz + phi * mu * u / B_g\n",
+    "\n",
+    "    # Conservation of energy\n",
+    "    conv_term = (4 / D_h) * h_coeff * (T_wall - T) * (2 * np.pi * R)\n",
+    "    chem_term = np.sum(hk_g * (phi * wdot_g + As * sdot_g))\n",
+    "\n",
+    "    res[offset_T] = dTdz  # isothermal condition\n",
```

Comment:
Shouldn't this be zero for the isothermal condition?

Reply:
Yes. It is indeed zero. But instead of specifying `res = 0`, the more stable version is `res = dTdz` to avoid the failure in Jacobian calculation.

Reply:
Ideally, `dtdz = 0` for isothermal conditions. But since we are including temperature `T` as the dependent variable, the residual equation is `res = dTdz - 0.0`  If we just hard code it to 0.0, then the solver fails due to singular Jacobian. We can make it an algebraic constraint equation instead of a differential eq. by setting it to 0.0. But then it means writing two different functions for isothermal and non-isothermal conditions.

Reply:
Great, thanks for the explanation! Can you add it to the notebook? 😁

Reply:
Done!